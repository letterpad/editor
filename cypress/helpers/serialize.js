import { parseFragment, serialize } from "parse5";

function removeData(root) {
  if (root.tagName === "iframe") {
    root.childNodes = [];
  }
  if (root.attrs) {
    root.attrs = root.attrs.filter(
      attr => !attr.name.startsWith("data-") && attr.name !== "class"
    );
  }
  if (root.childNodes) {
    for (const node of root.childNodes) {
      removeData(node);
    }
  }
}

export async function getHtmlContents() {
  return new Promise((resolve, reject) => {
    cy.get("div[contenteditable='true']").then(handle => {
      const fragment = parseFragment(handle[0].innerHTML);
      removeData(fragment);

      resolve(serialize(fragment));
    });
  });
}

export async function textContent(handle) {
  const textContent = await handle.getProperty("textContent");
  return textContent.jsonValue();
}
