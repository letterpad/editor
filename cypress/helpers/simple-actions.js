require("cypress-xpath");

export async function repeatKey(key, times) {
  for (let i = 0; i < times; i++) {
    await cy.focused().type(key);
  }
}

export async function clearEditor() {
  return new Promise((resolve, reject) => {
    cy.window().then(win => {
      win.__letterpadEditor.moveToRangeOfDocument();
      cy.focused()
        .type("{backspace}", { force: true })
        .then(resolve);
    });
  });
}

export async function clickXPath(xPath) {
  const ele = cy.xpath(xPath);
  if (ele.length > 0) {
    ele.click({ multiple: true });
  } else {
    ele.click();
  }
}
