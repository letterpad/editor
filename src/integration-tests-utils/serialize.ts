import { ElementHandle } from "puppeteer";
import cheerio from "cheerio";

export type EditorHandle = ElementHandle<Element>;

export async function getHtmlContents(handle: EditorHandle): Promise<string> {
  const context = await handle.executionContext();
  const html = await context.evaluate(node => node.innerHTML, handle);

  const cHtml = cheerio.load(html);
  const allNodes = cHtml("*");
  for (let i = 0; i < allNodes.length; i++) {
    const node = allNodes[i];
    for (let attr in node.attribs) {
      if (attr.startsWith("data-")) {
        allNodes.removeAttr(attr);
      }
    }
  }

  return cHtml.html();
}

export async function textContent(handle: EditorHandle): Promise<any> {
  const textContent = await handle.getProperty("textContent");
  return textContent.jsonValue();
}
