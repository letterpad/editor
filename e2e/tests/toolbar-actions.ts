import { EditorHandle, getHtmlContents } from "../serialize";
import { clearEditor, clickXPath } from "../simple-actions";

import params from "../params";

describe("features", () => {
  let editorHandle: EditorHandle;

  beforeAll(async () => {
    await page.goto(params.testServer);
    await page.waitForXPath("//div[@contenteditable='true']");
    const handle = await page.$('div[contenteditable="true"]');
    editorHandle = handle!;
    await editorHandle.focus();
  });

  describe("markdown", () => {
    beforeEach(async () => {
      await editorHandle.focus();
      await clearEditor(editorHandle);
      await page.evaluate(() => {
        window.prompt = () => "http://youtube.com/embed/link";
      });
      await page.evaluate(
        'document.querySelector("#letterpad-editor-toolbar-toggle-button").click()'
      );
      // await page.click("#letterpad-editor-toolbar-toggle-button");
      // await clickXPath("//*[@id='letterpad-editor-toolbar-toggle-button']");
      await page.waitFor(300);
    });

    test("audio", async () => {
      await clickXPath("//span[contains(text(), 'queue_music')]");
      const expected = await getHtmlContents(editorHandle);

      await clearEditor(editorHandle);
      const text = "[audio=http://youtube.com/embed/link]";
      await page.keyboard.type(text);
      // because this transformation adds an extra space
      await page.keyboard.press("Backspace");
      const actual = await getHtmlContents(editorHandle);

      expect(expected).toBe(actual);
    });

    test("embed", async () => {
      await clickXPath("//span[contains(text(), 'code')]");
      await page.waitFor(200);
      await page.keyboard.type("http://youtube.com/embed/link");
      await page.keyboard.press("Enter");
      const expected1 = await getHtmlContents(editorHandle);

      await clearEditor(editorHandle);
      const text = "[embed=http://youtube.com/embed/link]";
      await page.keyboard.type(text);
      // because this transformation adds an extra space
      await page.keyboard.press("Backspace");
      const expected2 = await getHtmlContents(editorHandle);

      expect(expected1).toBe(expected2);
    });

    test.only("image", async () => {
      await clickXPath("//span[contains(text(), 'image')]");
      await page.waitFor(200);
      await page.keyboard.type("image.jpg");
      await page.keyboard.press("Enter");
      const expected = await getHtmlContents(editorHandle);
      expect(expected).toMatchSnapshot();
    });

    test("separator", async () => {
      await clickXPath("//span[contains(text(), 'more_horiz')]");
      const actual = await getHtmlContents(editorHandle);
      expect(actual).toMatchSnapshot();
    });
  });
});
