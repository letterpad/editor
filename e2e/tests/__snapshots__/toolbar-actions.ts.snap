// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features markdown image 1`] = `"<html><head></head><body><div style=\\"position: relative;\\"><span><span><span>&#xFEFF;</span></span></span><span contenteditable=\\"false\\"><span style=\\"height: 0px; color: transparent; outline: none; position: absolute;\\"><span><span><span>&#xFEFF;</span></span></span></span><span contenteditable=\\"false\\"><div><img src=\\"http://localhost/embed/link\\"></div></span></span><span><span><span>&#xFEFF;<br></span></span></span></div><div style=\\"position: relative;\\"><span><span><span>&#xFEFF;<br></span></span></span></div></body></html>"`;

exports[`features markdown separator 1`] = `"<html><head></head><body><div style=\\"position: relative;\\"><span><span><span>&#xFEFF;<br></span></span></span></div><span style=\\"border-bottom: 2px solid rgb(0, 0, 0); display: block; opacity: 0.2;\\"><span><span><span>&#xFEFF;<br></span></span></span></span><div style=\\"position: relative;\\"><span><span><span>&#xFEFF;<br></span></span></span></div><div style=\\"position: relative;\\"><span><span><span>&#xFEFF;<br></span></span></span></div></body></html>"`;

exports[`features markdown youtube 1`] = `"<html><head></head><body><iframe type=\\"text/html\\" width=\\"640\\" height=\\"360\\" frameborder=\\"0\\" src=\\"http://localhost/embed/link\\"><span data-key=\\"259\\"><span data-slate-leaf=\\"true\\" data-offset-key=\\"259:0\\"><span data-slate-zero-width=\\"n\\" data-slate-length=\\"0\\">﻿<br></span></span></span></iframe><div style=\\"position: relative;\\"><span><span><span>&#xFEFF;<br></span></span></span></div></body></html>"`;

exports[`features markdown youtube 2`] = `"<html><head></head><body><iframe type=\\"text/html\\" width=\\"640\\" height=\\"360\\" frameborder=\\"0\\" src=\\"http://localhost/embed/link\\"><span data-key=\\"283\\"><span data-slate-leaf=\\"true\\" data-offset-key=\\"283:0\\"><span data-slate-zero-width=\\"n\\" data-slate-length=\\"0\\">﻿<br></span></span></span></iframe><div style=\\"position: relative;\\"><span><span><span>&#xFEFF;<br></span></span></span></div></body></html>"`;
